libs/SalesforceSDKCore/SalesforceSDKCore/Classes/IDP/SFSDKIDPAuthClient.m

```diff
@@ -0,0 +1,414 @@
+@end
+
+@implementation SFSDKIDPAuthClient
+@dynamic config;
+@dynamic context;
+@dynamic loginFlowSelectionController;
+
+- (instancetype) initWithConfig:(SFSDKOAuthClientConfig *)config {
+    self = [super initWithConfig:config];
+    return self;
+}
+
+- (SFSDKWindowContainer *) authWindow {
+    return [SFSDKWindowManager sharedManager].authWindow;
+}
+
+- (SFIDPLoginFlowSelectionCreationBlock)idpLoginFlowSelectionBlock {
+    if (!self.config.idpLoginFlowSelectionBlock) {
+        return ^SFSDKLoginFlowSelectionViewController * {
+            SFSDKLoginFlowSelectionViewController *controller = [[SFSDKLoginFlowSelectionViewController alloc] initWithNibName:nil bundle:nil];
+            controller.selectionFlowDelegate = self;
+            return controller;
+
+        };
+    }
+    return self.config.idpLoginFlowSelectionBlock;
+}
+
+- (SFIDPUserSelectionBlock)idpUserSelectionBlock {
+    if (!self.config.idpUserSelectionBlock) {
+        return ^SFSDKUserSelectionNavViewController * {
+            SFSDKUserSelectionNavViewController *controller = [[SFSDKUserSelectionNavViewController alloc] initWithNibName:nil bundle:nil];
+             return controller;
+        };
+    }
+    return self.config.idpUserSelectionBlock;
+}
+
+- (void)setIdpUserSelectionBlock:(SFIDPUserSelectionBlock)userSelectionBlock {
+    if (userSelectionBlock) {
+        self.config.idpUserSelectionBlock = userSelectionBlock;
+    }
+}
+
+- (BOOL)refreshCredentials:(void (^) (SFOAuthInfo *, SFUserAccount *))successCallbackBlock failure:(void (^)(SFOAuthInfo *, NSError *))failureBlock {
+    self.config.successCallbackBlock = successCallbackBlock;
+    self.config.failureCallbackBlock = failureBlock;
+    return [self refreshCredentials];
+}
+
+- (BOOL)refreshCredentials {
+    
+    if (self.config.isIdentityProvider) {
+        return [super refreshCredentials];
+    }
+    
+    if (self.credentials.accessToken==nil) {
+        if (self.context.userHint==nil) {
+            UIViewController<SFSDKLoginFlowSelectionView> *controller  = self.idpLoginFlowSelectionBlock();
+            controller.selectionFlowDelegate = self;
+            self.authWindow.viewController = controller;
+             [self.authWindow enable:YES withCompletion:nil];
+        } else {
+            [self launchIDPApp];
+        }
+    } else {
+        [super refreshCredentials];
+    }
+
+    return YES;
+}
+
+- (void)launchSPAppWithError:(NSError *)error reason:(NSString *)reason {
+    
+    NSURL *url = [self spAppURLWithError:error reason:reason];
+    [self.authWindow disable];
+    [[SFSDKWindowManager sharedManager].mainWindow enable];
+    BOOL launched  = [SFApplicationHelper openURL:url];
+    if (launched) {
+        if ( [self.config.idpDelegate respondsToSelector:@selector(authClient:didSendRequestForIDPAuth:)]) {
+            [self.config.idpDelegate authClient:self didSendRequestForIDPAuth:url];
+        }
+    }
+    
+}
+
+#pragma mark - private
+
+- (void)launchIDPApp {
```

Comment:
launchIDPApp calls invokeIDPApp on the main thread. Should launchSPApp do the same (do we need an invokeSPApp)?